### PR TITLE
[v1.13.x] prov/efa: back port RNR and AV fixes

### PR DESCRIPTION
**Pull request by wzamazon**
This PR back port recent bug fixes for RNR and address vector. Some commits are refactors, but is needed by the following bug fix commits therefore was pulled in too.

### CODE REVIEW

**Comment by wzamazon** [comment]
bot:aws:retest

**Comment by wzamazon** [comment]
bot:aws:retest

**Comment by wzamazon** [comment]
bot:aws:retest

**Comment by rwespetal** [comment]
> Some commits are refactors

Are there any new commits here or are all of these already in main?

**Comment by wzamazon** [comment]
All commits are already in main.

In the commit message, there are original commit ID it cherry-picked from.

**Comment by rwespetal** [comment]
Got it, thank you.

**Comment by rajachan** [comment]
0813fb3b changes FI_ERRNO_MAX in a public header. Do we want to change that within a release series?

**Comment by rwespetal** [comment]
> 0813fb3 changes FI_ERRNO_MAX in a public header. Do we want to change that within a release series?

Probably not, no. Wei, can you drop those commits so we can get the rest merged?

**Comment by wzamazon** [comment]
@rwespetal @rajachan Removed the last two commits

**Comment by rajachan** [comment]
👍

File: prov/efa/src/efa_av.c [hunk 1]
```diff
@@ -728,6 +728,31 @@ static int efa_av_lookup(struct fid_av *av_fid, fi_addr_t fi_addr,
 	return 0;
 }
 
+/*
+ * @brief remove a set of addresses from AV and release its resources
+ *
+ * This function implements fi_av_remove() for EFA provider.
+ *
+ * Note that even after an address was removed from AV, it is still
+ * possible to get TX and RX completion for the address. Per libfabric
```
Comment (rwespetal) [inline prov/efa/src/efa_av.c:737]: This isn't true - the behavior is undefined. If we can handle av remove with in flight ops by marking tx pkts in progress with FI_ADDR_NOTAVAIL/with connid in all headers so be it, but I do think we have more flexibility here in the av remove path than this commit suggests.

**Review by rwespetal — COMMENTED** [review]
Looks like the ENORX commits were squashed into the last commit, that needs to be fixed.

Other comments can be ignored/addressed elsewhere (this is the first time I'm reviewing these).

**Comment by wzamazon** [comment]
@rwespetal Fixed the last commit! Sorry about that.

**Review by rwespetal — APPROVED** [review]

File: prov/efa/src/rxr/rxr.h [hunk 2]
```diff
@@ -446,11 +446,12 @@ struct rxr_rx_entry {
 
 	/* linked with rx_entry_list in rdm_peer */
 	struct dlist_entry peer_entry;
+
+	/* linked with rx_entry_list in rxr_ep */
+	struct dlist_entry ep_entry;
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/rxr/rxr.h:452]: But bufpool did not provide a mechanism to loop over its entreis?

File: prov/efa/src/efa_av.c [hunk 3]
```diff
@@ -93,6 +93,7 @@ void efa_rdm_peer_init(struct rdm_peer *peer, struct rxr_ep *ep, struct efa_conn
 	ofi_recvwin_buf_alloc(&peer->robuf, rxr_env.recvwin_size);
 	peer->rx_credits = rxr_env.rx_window_size;
 	peer->tx_credits = rxr_env.tx_max_credits;
+	dlist_init(&peer->outstanding_tx_pkts);
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/efa_av.c:96]: Similiarly, ofi_bufpool did not provide a mechanism to loop over its entries. Maybe it's worth the effort to add such a function to ofi_bufpool.

[the remark this answers is not in the document]
Comment (rwespetal) [reply prov/efa/src/efa_av.c:96]: Yup, absolutely, we'll have to decide if we want to use something else or add that. Anyway, just thinking out loud at this point, what's here is fine.

[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/efa_av.c:96]: 👍 Agreed. We need to revisit this in the future.

**Pull request merged**